### PR TITLE
Implement Transaction Wrapper Model and Multi-Currency Support (Issue #16)

### PR DESCRIPTION
**Pull request by JuliaMathias**
### Overview
This PR introduces the `Moolah.Finance.Transaction` model, which provides a high-level, user-friendly interface for financial operations (Debits, Credits, and Transfers) while maintaining the integrity of the underlying double-entry `Moolah.Ledger` system. It also implements robust multi-currency support using the "Trading Account" pattern.

#### 1. Transaction Wrapper Model
- Added `Moolah.Finance.Transaction` resource with support for `:debit`, `:credit`, and `:transfer` types.
- Implemented automatic translation to ledger transfers via `CreateUnderlyingTransfer` and `UpdateUnderlyingTransfer` resource changes.
- Multi-currency support: Transactions can now track `source_amount` and `exchange_rate`, enabling transfers between accounts of different currencies (e.g., BRL -> USD).

#### 2. Trading Account Pattern (Multi-Currency Bridge)
- Implemented a system-managed "Trading Account" strategy.
- A multi-currency transfer is automatically split into two legs:
  1. `Source Account -> Trading Account` (Source Currency)
  2. `Trading Account -> Target Account` (Target Currency)
- This ensures that individual user accounts remain single-currency while allowing seamless exchanges through bridge accounts.

#### 3. Currency Integrity Validation
- Introduced a custom `CurrencyMatch` validation to enforce ledger safety:
  - **Single-Currency**: Ensures the transaction currency matches the account currency.
  - **Multi-Currency**: Ensures the `source_amount` matches the source account and the `amount` matches the target account.

#### 4. Atomicity and Real-time Updates
- Guaranteed transaction atomicity: Ledger transfers are created/updated within PostgreSQL transactions.
- Resolved "missed notifications" by explicitly capturing and propagating Ash notifications in `after_action` hooks, ensuring real-time UI updates work correctly.

### Technical Changes
- **Resources**: Created `Transaction`, updated `Account`, `Balance`, and `Transfer`.
- **Services**: Created `VirtualAccountService` to manage category-based and trading accounts.
- **Migrations**: Consolidated transaction fields and multi-currency support into two clean migrations.
- **Quality**: Resolved all Credo strict warnings and optimized code for readability (reduced function nesting).

### Verification
- **13 Automated Tests**: Covers all 6 original scenarios plus multi-currency updates and currency validation edge cases.
- **Precision**: Verified with today's exchange rates (e.g., 5.1972) to ensure zero rounding errors.
- **Clean Output**: Tests pass with zero failures and zero warnings.

---
*(Note: The tagging system has been deferred to a separate branch as requested.)*

### CODE REVIEW

File: lib/moolah/finance/resources/transaction.ex [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        :source_amount,
+        :description,
+        :date,
+        :budget_category_id,
+        :life_area_category_id
+      ]
+
+      change Moolah.Finance.Changes.UpdateUnderlyingTransfer
+      require_atomic? false
+    end
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/resources/transaction.ex]: The :update action uses a custom resource change (UpdateUnderlyingTransfer) that relies on before_transaction and after_action hooks to orchestrate the swapping of ledger legs. Since these hooks involve side effects that cannot be executed as a single atomic SQL statement, Ash requires require_atomic? false to validate the resource.

To address your concern about partial state and safety, I have explicitly added transaction? true to the action. This ensures that the entire sequence—creating new transfers, updating the transaction record, and destroying old legs—is wrapped in a database transaction and will roll back completely if any part fails. I've also added a strictly positive numericality validation to prevent any potential division by zero during exchange rate calculation

File: lib/moolah/finance/resources/transaction.ex [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  actions do
+    defaults [:read, :destroy]
+
+    create :create do
+      accept [
+        :transaction_type,
+        :amount,
+        :source_amount,
+        :description,
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/resources/transaction.ex]: I've addressed this by adding an absent(:source_amount) validation to the Transaction resource. This validation is conditionally enforced only when the transaction_type is not :transfer. I have also added a dedicated test case in 
transaction_test.exs to ensure that any attempt to include a source_amount in non-transfer actions results in a clear validation error: 'Source Amount can only be set for Transfer transactions'.

File: lib/moolah/finance/services/virtual_account_service.ex [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  ## key_prefix
+  - :expense -> "expense"
+  - :income -> "income"
+  """
+
+  @spec get_or_create(Ecto.UUID.t(), :expense | :income, String.t()) ::
+          {:ok, Ash.Resource.record()} | {:error, any()}
+  def get_or_create(category_id, key_prefix, currency \\ "BRL") do
+    # Construct a stable identifier using the UUID
+    identifier = "#{key_prefix}:#{category_id}"
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/services/virtual_account_service.ex]: I've resolved both the race condition and the currency multiplicity issue in VirtualAccountService.

- To fix the race condition, I've replaced the 'read-then-create' pattern with a declarative Ash upsert (upsert?: true) keyed on the :unique_identifier identity.
- To allow categories to support multiple currencies, I've updated the virtual account identifier to include the currency code (e.g., expense:USD:uuid). This ensures that each category has its own independent virtual account for every currency it interacts with."

File: lib/moolah/finance/changes/create_underlying_transfer.ex [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    account_id = Changeset.get_attribute(changeset, :account_id)
+    target_account_id = Changeset.get_attribute(changeset, :target_account_id)
+
+    budget_category_id = Changeset.get_attribute(changeset, :budget_category_id)
+    life_area_category_id = Changeset.get_attribute(changeset, :life_area_category_id)
+
+    case type do
+      :debit ->
+        # Debits use Budget Category for the virtual account
+        create_debit_transfer(account_id, budget_category_id, source_amount, source_currency)
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/changes/create_underlying_transfer.ex]: :debit and :credit transactions now strictly use the base amount, ensuring source_amount (multi-currency only) cannot leak into single-currency flows.

File: lib/moolah/finance/changes/update_underlying_transfer.ex [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            {:error, error} -> {:error, error}
+          end
+        end)
+
+      {:error, error} ->
+        Changeset.add_error(changeset, error)
+    end
+  end
+
+  @spec destroy_and_get_notifications(Ecto.UUID.t() | nil) ::
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/changes/update_underlying_transfer.ex]: Corrected ULID typespecs and updated Transfer documentation to clarify that the :destroy action is restricted to internal consistency corrections, maintaining the spirit of ledger immutability."

File: lib/moolah/finance/changes/create_underlying_transfer.ex [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+          {:ok, Ash.Resource.record() | map(), [Ash.Notifier.Notification.t()]} | {:error, any()}
+  def create_transfer_for_transaction(changeset) do
+    type = Changeset.get_attribute(changeset, :transaction_type)
+    amount_money = Changeset.get_attribute(changeset, :amount)
+    source_amount_money = Changeset.get_attribute(changeset, :source_amount)
+
+    # Extract amount and currency
+    amount = amount_money.amount
+    currency = amount_money.currency
+
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/changes/create_underlying_transfer.ex]: https://github.com/JuliaMathias/moolah/pull/53/commits/79c0f3bfa5dd7f0e2846251f061a44b12c11980e
